Sources/FirebaseCore/FutureProtocol.swift

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: BSD-3-Clause
+
+@_exported
+import firebase
+
+import Foundation
+
+@_spi(Internal)
+public typealias FutureCompletionType = swift_firebase.swift_cxx_shims.firebase.FutureCompletionType
+
+// This protocol enables extracting common code for Future handling. Because
+// C++ interop is limited for templated class types, we need to invent a
+// protocol to reflect the features of a Future<R> that should be generically
+// available. This works by having a C++ annotation (see ConformingFuture<R>)
+// that specifies this protocol conformance.
+@_spi(Internal)
+public protocol FutureProtocol {
+  associatedtype ResultType
+  func error() -> Int32
+  func __error_messageUnsafe() -> UnsafePointer<CChar>?
+  func __resultUnsafe() -> UnsafePointer<ResultType>?
+  func CallOnCompletion(_ completion: FutureCompletionType, _ user_data: UnsafeMutableRawPointer?)
+}
+
+// A home for various helper functions that make it easier to work with
+// Firebase Future objects from Swift.
+@_spi(Internal)
+public extension FutureProtocol {
+  // Callsites retain their own reference to the Future<R>, but they still need
+  // a way to know when the Future completes. This provides that mechanism.
+  // While the underlying Firebase `OnCompletion` method can provide a reference
+  // back to the Future, we don't need to expose that here.
+  func setCompletion(_ completion: @escaping () -> Void) {
+    CallOnCompletion({ ptr in
+      Unmanaged<CompletionWrapper>.fromOpaque(ptr!).takeRetainedValue().completion()
```

Comment:
Isn't force unwrapping basically equivalent? Won't it also result in a fatal error?

Reply:
Fair, but this will give you an error message that can helpful. I'll leave this to your discretion.

Reply:
I think the explicit fatalError is gratuitous here since it is just validating input from within this function. If we were building API that other code calls, then a more descriptive message (describing bad input) would make sense. Here we're really just asserting that `Unmanaged.passRetained...toOpaqueValue()` works as expected, so the `!` feels plenty sufficient.